aws/aws-parallelcluster-cookbook
# [develop] structured logging for compute node Chef recipe error

### Description of changes
* Created a Chef handler (`WriteComputeFleetSlurmChefError`) for compute nodes of Slurm scheduler that can create a log file (`bootstrap_error_msg`) to store Chef error related information into json structure
* Added this log file into CloudWatch agent configuration so it can be pushed to CloudWatch (only for compute nodes)
* Make minor changes to the [previous Chef handler](https://github.com/aws/aws-parallelcluster-cookbook/pull/1596):
  * move the code from the global entry point to within the config cookbook
  * only conditionally register this handler for head node bootstrap 
  * rename to `WriteHeadNodeChefError`

### Tests
Manual test:
* Introduce some errors in Chef recipe compute fleet-specific part, create a cluster using Slurm scheduler so that Chef failure will be triggered in the compute node during cluster creation
  * see the error file log in CloudWatch with expected logstream and json structure content
  * confirm in the cloud-init-output log, only the slurm compute node handler was registered and executed
* Introduce some errors in Chef recipe head node-specific part, create a cluster so that Chef failure will be triggered in the head node during cluster creation
  * do not see the error file log in CloudWatch log group
  * confirm in the chef-client log, only the head node handler was registered and executed


### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
